[Inductor] Enable the decomposition of quant/dequant per channel

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119177

**Summary**
Part 2 of fixing https://github.com/pytorch/pytorch/issues/119141 which needs vectorized code generation of per channel quant and int8 data type.
Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8_bf16_input
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8_bf16_input
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

leslie-fang-intel committed:
Enable the decomposition of quant/dequant per channel

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Fixing https://github.com/pytorch/pytorch/issues/119141. Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Fixing https://github.com/pytorch/pytorch/issues/119141. Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Fixing https://github.com/pytorch/pytorch/issues/119141. Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Fixing https://github.com/pytorch/pytorch/issues/119141. Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Fixing https://github.com/pytorch/pytorch/issues/119141. Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Fixing https://github.com/pytorch/pytorch/issues/119141. Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -1064,6 +1066,87 @@ def inner_fn(idx):
+        input.get_size()
+    ), f"Expecting axis to be < {len(input.get_size())}"
+
+    input, permute_axis_list = _permute_to_axis_zero(input, axis)
+    original_input_size = input.get_size()
+    input = view(input, (original_input_size[0], -1))
+    res = maximum(
+        quant_min,
+        minimum(
+            quant_max,
+            add(
+                round(mul(input, unsqueeze(div(1.0, scales), 1))),
+                unsqueeze(zero_points, 1),
+            ),
+        ),
+    )
+    res = view(res, original_input_size)
+    out = permute(res, tuple(permute_axis_list))
+    return to_dtype(out, dtype)
```
NIT: I think this would be more readable as a manually constructed `inner_fn`, where all this permuting and unsqueezing isn't necessary:
```
def inner_fn(idx):
    channel_idx = (idx[axis],)

    zero_point = zero_points_loader(channel_idx)
    inv_scale = ops.reciprical(scales_loader(channel_idx))

    val = ops.round(input_loader(idx) * inv_scale) + zero_point
    qmin, qmax = _create_constants(quant_min, quant_max, torch.float32)
    clamped = ops.max(qmin, ops.min(qmax, val))
    return ops.to_dtype(clamped, dtype)
   
```

Same principle applies to the dequantize operator.

leslie-fang-intel replied:
Thanks for the suggestion and changed, @peterbell10 please help to take a look again.

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -1064,6 +1066,87 @@ def inner_fn(idx):
+    assert len(zero_points.get_size()) == 1, "expect zero_points 1 dim"
+
+    if input.get_dtype() == torch.bfloat16:
+        input = input.to(torch.float32)
```
Should probably have a test for this case, also is `float16` relevant here?

jgong5 replied:
We only support `bfloat16` mixed quantization so far. `float16` will be supported in the future.

leslie-fang-intel replied:
Sorry for the mistake. Fix the code and add the bf16 input testcases. Please kindly help to take a look again.

Commit:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Part 2 of fixing https://github.com/pytorch/pytorch/issues/119141 which needs vectorized code generation of per channel quant and int8 data type.
Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Part 2 of fixing https://github.com/pytorch/pytorch/issues/119141 which needs vectorized code generation of per channel quant and int8 data type.
Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Part 2 of fixing https://github.com/pytorch/pytorch/issues/119141 which needs vectorized code generation of per channel quant and int8 data type.
Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] Enable the decomposition of quant/dequant per channel"


**Summary**
Part 2 of fixing https://github.com/pytorch/pytorch/issues/119141 which needs vectorized code generation of per channel quant and int8 data type.
Enable decomposition of quant/dequant per channel to make it vectorized code generation.

**TestPlan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_uint8
python -u -m pytest -s -v test_cpu_repro.py -k test_per_channel_fake_quant_int8
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1064,6 +1066,96 @@ def inner_fn(idx):
+    assert len(scales.get_size()) == 1, "expect scales 1 dim"
+    assert len(zero_points.get_size()) == 1, "expect zero_points 1 dim"
+
+    if input.get_dtype() == torch.bfloat16:
```
torch.float16 not handled on intentionally?

leslie-fang-intel replied:
Yes, since we only support `bfloat16` mixed quantization so far, `float16` will be supported in the future.

leslie-fang-intel commented:
@pytorchbot merge
